test/end2end_test.go

```diff
@@ -1333,6 +1333,46 @@ func testConcurrentServerStopAndGoAway(t *testing.T, e env) {
 	awaitNewConnLogOutput()
 }
 
+func (s) TestDetailedConnectionCloseErrorPropagatesToRpcError(t *testing.T) {
+	for _, e := range listTestEnv() {
+		testDetailedConnectionCloseErrorPropagatesToRPCError(t, e)
+	}
+}
+
+func testDetailedConnectionCloseErrorPropagatesToRPCError(t *testing.T, e env) {
+	te := newTest(t, e)
+	te.userAgent = testAppUA
+	rpcStartedOnServer := make(chan struct{})
+	rpcDoneOnClient := make(chan struct{})
+	te.streamServerInt = func(srv interface{}, ss grpc.ServerStream, info *grpc.StreamServerInfo, handler grpc.StreamHandler) error {
+		close(rpcStartedOnServer)
+		<-rpcDoneOnClient
+		return status.Error(codes.Internal, "arbitrary status")
+	}
+	te.startServer(&testServer{security: e.security})
+	defer te.tearDown()
+
+	cc := te.clientConn()
+	tc := testpb.NewTestServiceClient(cc)
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	possibleConnResetMsg := "connection reset by peer"
```

Comment:
When and how would this error occur?

Reply:
Good question. I think I actually left this in while I was developing the test and before I had the `rpcStartedOnServer` channel to wait for the server handler to receive the RPC (preventing us from stopping the server before headers had been sent from client to server). When we remove that channel and the synchronization it provides, stopping the server can easily race with the client's attempt to send RPC headers to the server, and we can see frequent RST packets being sent from server to client in this case (I guess, if a socket closes while they're still un-processed data in the socket buffer, then the kernel will generate an RST to send back to the peer) - in this case, the error message will be "connection reset by peer".

That said, I think that the "synchronization" provided by the `rpcStartedOnServer` channel is fundamentally brittle, because AFAICT the client can still e.g. send a BDP ping to the server - i.e., this channel doesn't give us an actual guarantee that no more data will be sent from to the client, so I think the test is actually more robust if we remove this channel entirely.